File: src/main/java/com/baulsupp/oksocial/services/squareup/SquareUpAuthFlow.java
package com.baulsupp.oksocial.services.squareup;

import com.baulsupp.oksocial.authenticator.AuthUtil;
import com.baulsupp.oksocial.authenticator.SimpleWebServer;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token;
import ee.schimke.oksocial.output.OutputHandler;
import ee.schimke.oksocial.output.util.JsonUtil;
import java.io.IOException;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;

import static java.util.stream.Collectors.joining;

public class SquareUpAuthFlow {

  public static Oauth2Token login(OkHttpClient client, OutputHandler outputHandler, String clientId,
      String clientSecret, Set<String> scopes) throws IOException {
    try (SimpleWebServer<String> s = SimpleWebServer.forCode()) {

      String serverUri = s.getRedirectUri();

      String loginUrl = "https://connect.squareup.com/oauth2/authorize"
          + "?client_id=" + clientId
          + "&redirect_uri=" + URLEncoder.encode(serverUri, "UTF-8")
          + "&response_type=code"
          + "&scope=" + URLEncoder.encode(scopes.stream().collect(joining(" ")),
          "UTF-8");

      outputHandler.openLink(loginUrl);

      String code = s.waitForCode();

      String tokenUrl = "https://connect.squareup.com/oauth2/token";
      Map<String, String> map = new HashMap<>();
      map.put("client_id", clientId);
      map.put("client_secret", clientSecret);
      map.put("code", code);
      map.put("redirect_uri", serverUri);
      String body = JsonUtil.toJson(map);

      RequestBody reqBody = RequestBody.create(MediaType.parse("application/json"), body);
      Request request = new Request.Builder().url(tokenUrl).post(reqBody).build();
      Map<String, Object> responseMap = AuthUtil.makeJsonMapRequest(client, request);

      return new Oauth2Token((String) responseMap.get("access_token"));
    }
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/squareup/SquareUpAuthFlow.kt
package com.baulsupp.oksocial.services.squareup

import com.baulsupp.oksocial.authenticator.SimpleWebServer
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2TokenResponse
import com.baulsupp.oksocial.credentials.NoToken
import com.baulsupp.oksocial.kotlin.postJsonBody
import com.baulsupp.oksocial.kotlin.query
import com.baulsupp.oksocial.kotlin.request
import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.oksocial.services.squareup.model.AuthDetails
import okhttp3.OkHttpClient
import okhttp3.Response
import java.net.URLEncoder.encode

object SquareUpAuthFlow {

  suspend fun login(client: OkHttpClient, outputHandler: OutputHandler<Response>, clientId: String,
                    clientSecret: String, scopes: Iterable<String>): Oauth2Token {
    SimpleWebServer.forCode().use { s ->
      val loginUrl = "https://connect.squareup.com/oauth2/authorize?client_id=$clientId&redirect_uri=${encode(s.redirectUri, "UTF-8")}&response_type=code&scope=${encode(scopes.joinToString(" "), "UTF-8")}"

      outputHandler.openLink(loginUrl)

      val code = s.waitForCode()

      val response = client.query<Oauth2TokenResponse>(request("https://connect.squareup.com/oauth2/token", NoToken) {
        postJsonBody(AuthDetails(clientId, clientSecret, code, s.redirectUri))
      })

      return Oauth2Token(response.access_token)
    }
  }
}
